### DISCUSSION TITLE
CONFIG_NO_HZ_FULL vs. CONFIG_NO_HZ_IDLE

### DISCUSSION THREAD

**Discussion by mabod** [Frogging-Family/linux-tkg#937] [General]
The linux-tkg default setting for tickless kernel is  `CONFIG_NO_HZ_IDLE=y`
This is what you get when you hit RETURN through all the linux-tkg config questions and let the script pick the defaults.

This is also the recommendation of the kernel developers:

> 2. Omit scheduling-clock ticks on idle CPUs (**CONFIG_NO_HZ_IDLE**=y or CONFIG_NO_HZ=y for older kernels). This is the most common approach, and **should be the default**.
> 
> 3. Omit scheduling-clock ticks on CPUs that are either idle or that have only one runnable task (**CONFIG_NO_HZ_FULL**=y). Unless you are running realtime applications or certain types of HPC workloads, **you will normally -not- want this option.**

(from: https://docs.kernel.org/timers/no_hz.html)

But the main distros like arch, debian and fedora all have `CONFIG_NO_HZ_FULL=y`
Any idea why they are doing that?

**Comment by mabod** [comment]
I found the answer  myself.  In 2021 a change was introduced to the kernel source which better explains what CONFIG_NO_HZ_FULL is actually doing.

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=176b8906c399a170886ea4bad5b24763c6713d61

It says:

```
CONFIG_NO_HZ_FULL:
...
By default, without passing the nohz_full parameter, this behaves just
like NO_HZ_IDLE.

If you're a distro say Y.
```

With that said I believe linux-tkg should also default to `CONFIG_NO_HZ_FULL=y`

**Reply by ptr1337** [reply]
If `RCU_NOCB_DEFAULT_ALL` is set, it will offload automatically the CPUs, even if no kernel parameters are set

**Reply by mabod** [reply]
What does your answer tell me? CONFIG_NO_HZ_FULL=y is the default in arch, debian and fedora. RCU_NOCB_DEFAULT_ALL is not set in arch.

I have opened an issue for this topic: https://github.com/Frogging-Family/linux-tkg/issues/938
